File: api/internal/database/redis.go
package database

import (
	"context"
	"os"
	"time"

	"github.com/redis/go-redis/v9"
	_ "github.com/joho/godotenv/autoload"
	"github.com/rs/zerolog/log"
)

var redisClient *redis.Client

// GetRedisClient returns a singleton Redis client
func GetRedisClient() *redis.Client {
	if redisClient != nil {
		return redisClient
	}

	redisURL := os.Getenv("REDIS_URL")
	if redisURL == "" {
		redisURL = "redis://localhost:6379"
	}

	opt, err := redis.ParseURL(redisURL)
	if err != nil {
		log.Fatal().Err(err).Msg("Failed to parse REDIS_URL")
	}

	redisClient = redis.NewClient(opt)

	// Test connection
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	if err := redisClient.Ping(ctx).Err(); err != nil {
		log.Fatal().Err(err).Msg("Failed to connect to Redis")
	}

	log.Info().Str("url", redisURL).Msg("Successfully connected to Redis")

	return redisClient
}

File: src/lib/server/redis.ts
import { createClient } from 'redis';
import { REDIS_URL } from '$env/static/private';

type RedisClient = ReturnType<typeof createClient>;

let client: RedisClient | null = null;

/**
 * Get or create Redis client singleton
 */
export async function getRedisClient(): Promise<RedisClient> {
	if (client) {
		return client;
	}

	client = createClient({
		url: REDIS_URL || 'redis://localhost:6379'
	});

	client.on('error', (err) => {
		console.error('Redis Client Error:', err);
	});

	client.on('connect', () => {
		console.log('Redis Client Connected');
	});

	await client.connect();

	return client;
}

/**
 * Gracefully disconnect Redis client
 */
export async function disconnectRedis(): Promise<void> {
	if (client) {
		await client.quit();
		client = null;
	}
}

/**
 * Check if Redis is available and connected
 */
export async function isRedisAvailable(): Promise<boolean> {
	try {
		if (!client) {
			await getRedisClient();
		}
		await client!.ping();
		return true;
	} catch (error) {
		console.error('Redis not available:', error);
		return false;
	}
}
